app/src/main/java/co/nimblehq/alpine/sample/nfc/NfcScanningActivity.kt

```diff
@@ -1,14 +1,148 @@
+            showMessage(getString(R.string.nfc_scanning_reading_data))
+            val passportInfo = withContext(Dispatchers.IO) { nfcReader.readNfc(tag, mrzInfo) }
+            if (passportInfo == null) {
+                showError()
+            } else {
+                loadPassportInfo(passportInfo)
+            }
+        }
+    }
+
+    private fun showError() {
+        binding.gPassportDetails.visibility = GONE
+        binding.pbLoading.visibility = GONE
+        binding.tvInstruction.visibility = VISIBLE
+        showMessage(getString(R.string.error_generic))
+    }
+
+    private fun loadPassportInfo(passportInfo: PassportInfo) {
+        with(passportInfo) {
+            with(binding) {
+                gPassportDetails.visibility = VISIBLE
+                pbLoading.visibility = GONE
+                personDetails?.run {
+                    tvName.text = getString(R.string.nfc_scanning_name, name)
+                    tvSurname.text = getString(R.string.nfc_scanning_surname, surname)
+                    tvPersonalNumber.text = getString(R.string.nfc_scanning_personal_number, personalNumber)
+                    tvGender.text = getString(R.string.nfc_scanning_gender, gender)
+                    tvBirthDate.text = getString(R.string.nfc_scanning_birth_date, birthDate)
+                    tvExpiryDate.text = getString(R.string.nfc_scanning_expiry_date, expiryDate)
+                    tvSerialNumber.text = getString(R.string.nfc_scanning_passport_number, documentNumber)
+                    tvNationality.text = getString(R.string.nfc_scanning_nationality, nationality)
+                    tvIssuerAuthority.text = getString(R.string.nfc_scanning_issuer_authority, issuingState)
+                }
+                tvDocumentType.text = getString(R.string.nfc_scanning_document_type, documentType?.name)
+                biometrics?.run {
+                    faceImage?.bitmap?.let { bitmap ->
+                        ivPassportImage.load(bitmap)
+                        gPassportImage.visibility = VISIBLE
+                    } ?: run { gPassportImage.visibility = GONE }
+                    signatureImage?.bitmap?.let { bitmap ->
+                        ivSignatureImage.load(bitmap)
+                        gSignatureImage.visibility = VISIBLE
+                    } ?: run { gSignatureImage.visibility = GONE }
+                    fingerprints?.firstOrNull()?.let { bitmap ->
+                        ivFingerprintImage.load(bitmap)
+                        gFingerprintImage.visibility = VISIBLE
+                    } ?: run { gFingerprintImage.visibility = GONE }
+                }
+            }
+        }
+    }
+
+    private fun showMessage(message: String) = Snackbar.make(binding.root, message, Snackbar.LENGTH_SHORT).show()
+
+    companion object {
+        private const val ARG_MRZ = "ARG_MRZ"
+        private const val NFC_TECH_ISO_DEP = "android.nfc.tech.IsoDep"
+
+        fun start(context: Context, mrzInfo: MrzInfo) {
```

Comment:
Forcing to pass the `Activity` instance is a better practice to start the activity. In case you pass the wrong context, `Application` for example, might result in an error.
```suggestion
        fun start(activity: Activity, mrzInfo: MrzInfo) {
```